From aeb54c24beffb07f912d7b7ed4a37150db4ecd31 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Tue, 14 Oct 2025 19:13:33 -0500
Subject: [PATCH 346/441] winewayland: Implement support for
 WINE_DISABLE_VULKAN_OPWR.

---
 dlls/winewayland.drv/vulkan.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/dlls/winewayland.drv/vulkan.c b/dlls/winewayland.drv/vulkan.c
index 3c47e3e4810..aa67378dc0f 100644
--- a/dlls/winewayland.drv/vulkan.c
+++ b/dlls/winewayland.drv/vulkan.c
@@ -66,6 +66,22 @@ static void wine_vk_surface_destroy(struct wayland_client_surface *client)
     if (data) wayland_win_data_release(data);
 }
 
+static BOOL vulkan_opwr_disabled(void)
+{
+    static int disabled = -1;
+
+    if (disabled == -1)
+    {
+        const char *env = getenv("WINE_DISABLE_VULKAN_OPWR");
+        if (env && !strcmp(env, "1"))
+            disabled = 1;
+        else
+            disabled = 0;
+    }
+
+    return disabled;
+}
+
 static VkResult wayland_vulkan_surface_create(HWND hwnd, const struct vulkan_instance *instance, VkSurfaceKHR *surface, void **private)
 {
     VkResult res;
@@ -84,6 +100,8 @@ static VkResult wayland_vulkan_surface_create(HWND hwnd, const struct vulkan_ins
     tid = NtUserGetWindowThread(hwnd, &pid);
     if (tid && pid != GetCurrentProcessId())
     {
+        if (vulkan_opwr_disabled()) return VK_ERROR_OUT_OF_HOST_MEMORY;
+
         ERR("Cross process rendering is not supported!\n");
         return VK_ERROR_OUT_OF_HOST_MEMORY;
     }
-- 
2.51.0

